mypyc/lib-rt/misc_ops.c

```diff
@@ -708,3 +708,72 @@ CPy_CallReverseOpMethod(PyObject *left,
     Py_DECREF(m);
     return result;
 }
+
+PyObject *CPySingledispatch_RegisterFunction(PyObject *singledispatch_func,
+                                        PyObject *cls,
+                                        PyObject *func) {
+    PyObject *registry = PyObject_GetAttrString(singledispatch_func, "registry");
+    PyObject *register_func = NULL;
+    PyObject *typing = NULL;
+    PyObject *get_type_hints = NULL;
+    PyObject *type_hints = NULL;
+
+    if (registry == NULL) goto fail;
+    if (func == NULL) {
+        // one argument case
+        if (PyType_Check(cls)) {
+            // passed a class
+            // bind cls to the first argument so that register gets called again with both the
+            // class and the function
+            register_func = PyObject_GetAttrString(singledispatch_func, "register");
```

Comment:
Need an error check here.

(The cases where it could error are silly: a caller reads register out, then `delattr`s it, or something. In general, I think, with the python C API, just never assume that something can't error.)